blog: table v9 reactivity (#1048)
TanStack/tanstack.com: src/blog/tanstack-table-v9-reactivity.md

---
title: Inside TanStack Table V9 Reactivity
published: 2026-07-19
excerpt: TanStack Table V9 moves state and options behind reactive primitives so updates stay within the scopes that consume them. Here is the path we took to get there.
library: table
authors:
  - Riccardo Perra
---

![TanStack Table V9 - Reactivity](/blog-assets/tanstack-table-v9-reactivity/header.png)

You might have found yourself in a situation where a data grid may have started as just a simple table, but then it slowly became a small application surface in of itself. One column renders a row-selection widget, another renders an inline editor, and a toolbar above the table shows how many rows are selected. When row selection changes, the selection widget and the counter need to update, while the other cells, editors, and pagination controls should not re-render just because one row became selected. Throw in features like search or especially column resizing, and you will find yourself in a situation where rendering performance really starts to matter.

> That rendering problem sits behind much of the recent TanStack Table V9 reactivity work. TanStack Table, along with an increasingly higher number of other TanStack libraries, challenges itself with having adapters for as many JavaScript frameworks as possible. Table V9 is shipping with a [record 9 dedicated framework adapters](https://tanstack.com/table/latest/docs/framework) so far, plus improved support for using it with no framework adapter at all in vanilla JavaScript. We have had to iterate multiple times during the development of Table V9 to get this correct — both in being performant and easy to maintain.

Reviewer: ```suggestion
## Rethinking Reactivity for Table V9

That rendering problem sits behind much of the recent TanStack Table V9 reactivity work. TanStack Table, along with an increasingly higher number of other TanStack libraries, challenges itself with having adapters for as many JavaScript frameworks as possible. Table V9 is shipping with a [record 9 dedicated framework adapters](https://tanstack.com/table/latest/docs/framework) so far, plus improved support for using it with no framework adapter at all in vanilla JavaScript. We have had to iterate multiple times during the development of Table V9 to get this correct — both in being performant and easy to maintain.
```

The broader [TanStack Table V9: Taking Form](https://tanstack.com/blog/tanstack-table-v9-taking-form) announcement introduced the new state management and rendering controls. Table V9 keeps most of the same public method calls while moving state and options into a reactive graph that each adapter can connect to its framework.

Users should not have to understand that graph in order to build a table. Even though TanStack Table is headless, one of our requirements for Table V9 was that it fit naturally into each framework's reactive model. The complexity of translating between the framework-agnostic core and each rendering model belongs inside the TanStack Table library, not in your application code.

> ## TL;DR

Reviewer: ```suggestion
## TL;DR: Where V9 Landed
```

TanStack Table V9 moves state and options behind reactive primitives so updates can stay within the scopes that consume them. Pagination, row selection, column sizing, filters, and the other feature states are managed by separate atoms, while `table.store` derives the aggregated state view. A Table method can depend on the state and options it reads without subscribing to the entire table-state snapshot.

> Table creates and manages those atoms through the adapter. Applications continue to pass ordinary values, including values read from framework signals or refs. Signal-native adapters implement the atom contract with their own primitives, while React uses TanStack Store and provides opt-in granular rendering through selectors and `Subscribe`.

Reviewer: ```suggestion
A Table computation should react to the state it actually reads. Selecting a row should update its checkbox and the selected-row counter without forcing every cell to depend on the entire table state.

TanStack Table V9 gets there by representing each feature-state slice (pagination, row selection, column sizing, filters, and others) with its own reactive atom, while `table.store` derives the aggregated state view for compatibility.

The core reactivity depends on a small shared atom contract provided by @tanstack/store:
- Signal-native adapters implement it with their framework primitives and can keep option reads in the same reactive graph. 
- Store-backed adapters keep options synchronized during render; in React, granular rendering is opt-in through selectors and `Subscribe`.

Applications still pass ordinary values and call the same Table methods. The rest of this article retraces how we got here: the adapter-level approaches we tried, the limits they exposed, and why the reactive boundary ultimately had to move underneath the Table APIs.
```

## The Table V8 Contract

> TanStack Table V8 used a contract that served the library well across frameworks. The core built one table instance, attached feature APIs to it, stored resolved options on it, and exposed `setOptions` and `setState` so each adapter could keep that instance aligned with its framework.

Reviewer: ```suggestion
## The Table V8 Starting Point

Before V9 moved reactivity underneath the Table APIs, adapters connected the stable V8 table instance to each framework's update model.

### Keeping a Stable Table Instance in Sync

TanStack Table V8 used a contract that served the library well across frameworks. The core built one table instance, attached feature APIs to it, stored resolved options on it, and exposed `setOptions` and `setState` so each adapter could keep that instance aligned with its framework.
```

Controlled pagination gives a concrete example: when a user clicks the next-page button, Table calls `onPaginationChange` and user code updates framework state. On the next render, the adapter passes the new `state.pagination` value back through `table.setOptions`, allowing the core to read `table.options.state.pagination` while continuing to use the same table instance.

```ts
const table = createTable({
  state: { pagination },
  onPaginationChange: setPagination,
  // ...
})

table.setOptions((prev) => ({
  ...prev,
  state: { ...prev.state, pagination },
}))
```

That contract kept the core portable, but the framework could not see which state, option, row, column, or cell value a call such as `table.getRowModel()` touched, even when Table's own memoization could reuse the result. Any connection between those reads and the framework render cycle had to be built by the adapter from the outside.

The React adapter also predated React Compiler, and the [Table V8 documentation](https://tanstack.com/table/v8/docs/installation) warned that compiler-enabled apps might not work.

> ## Making Table V8 Reactive In Angular

Reviewer: ```suggestion
### Making Table V8 Reactive In Angular with Signals
```

> That contract kept the core portable, but the framework could not see which state, option, row, column, or cell value a call such as `table.getRowModel()` touched, even when Table's own memoization could reuse the result. Any connection between those reads and the framework render cycle had to be built by the adapter from the outside.
> 
> The React adapter also predated React Compiler, and the [Table V8 documentation](https://tanstack.com/table/v8/docs/installation) warned that compiler-enabled apps might not work.
> 
> ## Making Table V8 Reactive In Angular
> 
> At the time, TanStack Table did not have an Angular adapter, and I was contributing to the project from the outside just as Angular introduced signals. The new reactive model made an Angular adapter practical to explore, and building it became my entry point into Table.
> 
> I kept the Table V8 core contract intact. `createAngularTable` accepted a reactive options callback, allowing the adapter to track values such as `data()` and synchronize the latest options through `setOptions`.
> 
> ```ts
> const data = signal<Person[]>([])
> 
> const table = createAngularTable(() => ({
>   data: data(),
>   columns,
> }))

Reviewer: ```suggestion
const data = signal<Person[]>([]);
const pagination = signal<PaginationState>({
  pageIndex: 0,
  pageSize: 20
})

const table = createAngularTable(() => ({
  data: data(),
  columns,
  state: {
    pagination: pagination()
  }
}))
```

In this case, `flexRender` could keep passing the same `CellContext` reference while the values returned by its Table methods changed. Angular saw neither a new input nor a signal dependency behind those methods, so the adapter had to coordinate table updates with view checks to keep the component current.

The Angular adapter went through several iterations to make ordinary Table calls update reliably, including components rendered through `flexRender` with stable inputs. Its reactive boundary remained broad, however: the proxy only covered methods reached through the table instance, while `flexRender` had to re-check rendered content broadly instead of following only the state read by each component.

React Compiler later exposed a similar boundary in React. A cell component could receive a stable `row` reference while `row.getIsSelected()` changed behind it, allowing the compiler to reuse memoized JSX because it could not see the state read hidden inside the method. Both cases pointed to the same Table V9 requirement: move reactivity underneath the Table APIs so rows, cells, headers, and columns can expose the dependencies they actually read.

> ## When Wrapping Methods Was Not Enough

Reviewer: ```suggestion
## Where V9 Reactivity Started: Wrapping Table Methods
```

That solved the proxy's limited reach, but attached reactive machinery to the API objects created for each table instance, including the rows and cells materialized by its row models. Some of that work could be deferred until first access, but constructing row models and rendering the table still exercised part of the API, while creating computed wrappers lazily complicated method access and caching. With `R` rows and `C` cells per row, the cell portion alone covered `R × C` objects. If each cell exposed `N` eligible methods, the wrapping surface grew roughly as `R × C × N`, before counting the methods on rows, headers, columns, and the table itself. Those methods ultimately read a much smaller set of shared state and options, yet reactivity followed the shape of the constructed object graph.

The boundary needed to move underneath the methods instead. Making state and option reads reactive would let the same APIs expose their dependencies without wrapping each method on every object.

> ## Store Moved State Into The Core

Reviewer: ```suggestion
## Moving the Reactive Boundary into the Core
```

TanStack Table first adopted [TanStack Store](https://github.com/TanStack/table/pull/6143) as part of the Table V9 rewrite. The initial implementation used the class-based `Store` and `Derived` APIs available at the time: one base store held Table-owned state, while a derived table-state snapshot reconciled it with controlled values such as `state.rowSelection`. State management had moved into the core, but every feature still read from one aggregated snapshot, while options remained plain values outside the store.

Shortly afterward, the [`@tanstack/store` 0.9 update](https://github.com/TanStack/table/pull/6180) moved Store itself to a signal-backed implementation built on a modified version of `alien-signals`. Table adopted the new `createStore` and `createAtom` API, bringing signal-based dependency tracking into the core without changing its high-level state model.

> ## Options Became Part Of The Graph

Reviewer: ```suggestion
## Bringing Table Options into the Reactive Model
```

The `enableRowSelection` option is a good example. A row checkbox depends on `rowSelection` state and on whether that row can be selected at all. If an Angular, Solid, or Vue app changes `enableRowSelection` from a reactive value, `row.getCanSelect()` needs to see that option change. Updating selection state alone cannot fix a stale option read.

[PR #6181](https://github.com/TanStack/table/pull/6181) added an `optionsStore`. Instead of storing resolved options only as a plain object on the table, the table could read options through a store-backed getter. `table.setOptions` updated that store, and table APIs that touched `table.options` could participate in the same reactive graph as state reads.

```ts
Object.defineProperty(table, 'options', {
  get() {
    return table.optionsStore.state
  },
  set(value) {
    table.optionsStore.setState(() => value)
  },
})
```

Options reactivity brought the core model closer to how users write table code, where `data`, feature flags, and callbacks come from components and need to remain live inputs rather than construction-time configuration.

> ## One Core, Two Reactive Graphs

Reviewer: ```suggestion
## Keeping Two Reactive Graphs in Sync
```

A simplified version of the page-selection calculation makes those dependencies visible:

```ts
function getIsAllPageRowsSelected(table) {
  const rows = table.getPaginatedRowModel().flatRows
  const rowSelection = table.atoms.rowSelection.get()
  const enableRowSelection = table.options.enableRowSelection

  const selectableRows = rows.filter((row) =>
    typeof enableRowSelection === 'function'
      ? enableRowSelection(row)
      : (enableRowSelection ?? true),
  )

  return (
    selectableRows.length > 0 &&
    selectableRows.every((row) => Boolean(rowSelection[row.id]))
  )
}
```

In a native signal adapter, evaluating this method inside a component, `computed`, or `createMemo` records those three incoming reads automatically.

Per-feature atoms gave the graph the granularity we wanted, leaving one question: did every adapter have to build those atoms with the same runtime implementation, or could the core depend only on their shared contract?

> ## The Shared Interface

Reviewer: ```suggestion
## The Final Design: One Contract, Many Reactive Runtimes
```

The core was already framework-agnostic in Table V8; Table V9 makes its reactivity layer portable too. The recently added [Ember adapter](https://github.com/TanStack/table/pull/6397), contributed by [Ryan Tablada](https://github.com/rtablada), shows that the contract does not require a conventional signals API: Ember's `@tracked` writable values and `@cached` derived values only need to expose compatible `Atom` and `ReadonlyAtom` surfaces. Table V9 also adds Preact and Alpine adapters, which provide Store-backed bindings. Rendering and lifecycle remain framework-specific, but a new adapter no longer needs a separate Table state model.

## What This Changes For Table Users

Reactivity now sits below the Table APIs, where each adapter can connect state and option reads to the component or subtree that should update.

> ### Native Signals Where They Exist

Reviewer: ```suggestion
### Automatic Tracking in Signal-Native Adapters
```
